### PR TITLE
Refactor function 'new_bitmap_index_iterator'

### PR DESCRIPTION
**Pull request by xlwh**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [ ] enhancement
- [x] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Refactor 'Status BitmapIndexReader::new_iterator(BitmapIndexIterator** iterator)' to 'StatusOr<std::unique_ptr<BitmapIndexIterator>> BitmapIndexReader::new_iterator()'.

### CODE REVIEW

File: be/src/storage/rowset/vectorized/segment_iterator.cpp [hunk 1]
```diff
@@ -1371,10 +1374,6 @@ void SegmentIterator::close() {
 
     STLClearObject(&_selection);
     STLClearObject(&_selected_idx);
-
-    for (auto* iter : _bitmap_index_iterators) {
```
Comment (imay) [inline be/src/storage/rowset/vectorized/segment_iterator.cpp:1375]: Better to clear _bitmap_index_iterators to free memory earlier.

**Review by imay — DISMISSED** [review]

File: be/src/storage/rowset/vectorized/segment_iterator.cpp [hunk 2]
```diff
@@ -1232,11 +1228,13 @@ Status SegmentIterator::_encode_to_global_id(ScanContext* ctx) {
 
 Status SegmentIterator::_init_bitmap_index_iterators() {
     DCHECK_EQ(_predicate_columns, _opts.predicates.size());
-    _bitmap_index_iterators.resize(ChunkHelper::max_column_id(_schema) + 1, nullptr);
+    _bitmap_index_iterators.reserve(ChunkHelper::max_column_id(_schema) + 1);
```
Comment (sduzh) [inline be/src/storage/rowset/vectorized/segment_iterator.cpp:1231]: It should be fine to remove this reservation.

File: be/src/storage/rowset/vectorized/segment_iterator.cpp [hunk 3]
```diff
@@ -1232,11 +1228,13 @@ Status SegmentIterator::_encode_to_global_id(ScanContext* ctx) {
...
     for (const auto& pair : _opts.predicates) {
-        ColumnId cid = pair.first;
-        if (_bitmap_index_iterators[cid] == nullptr) {
-            RETURN_IF_ERROR(_segment->new_bitmap_index_iterator(cid, &_bitmap_index_iterators[cid]));
+        const ColumnId cid = pair.first;
+        if (!ContainsKey(_bitmap_index_iterators, cid)) {
+            ASSIGN_OR_RETURN(std::unique_ptr<BitmapIndexIterator> bitmap_index_iterator,
+                             _segment->new_bitmap_index_iterator(cid));
+            _bitmap_index_iterators[cid] = std::move(bitmap_index_iterator);
             _has_bitmap_index |= (_bitmap_index_iterators[cid] != nullptr);
```
Comment (sduzh) [inline be/src/storage/rowset/vectorized/segment_iterator.cpp:1238]: We can now rewrite this line to `_has_bitmap_index = true`

File: be/src/storage/rowset/vectorized/segment_iterator.cpp [hunk 4]
```diff
@@ -235,7 +231,7 @@ class SegmentIterator final : public ChunkIterator {
     vectorized::SegmentReadOptions _opts;
     RawColumnIterators _column_iterators;
     ColumnDecoders _column_decoders;
-    std::vector<BitmapIndexIterator*> _bitmap_index_iterators;
+    std::unordered_map<ColumnId, std::unique_ptr<BitmapIndexIterator>> _bitmap_index_iterators;
```
Comment (sduzh) [inline be/src/storage/rowset/vectorized/segment_iterator.cpp:234]: It's good to replace `std::vector` with `std::unordered_map`, especially when the total number of columns is large, `std::unordered_map` may use less memory.

Comment (xlwh) [reply be/src/storage/rowset/vectorized/segment_iterator.cpp:234]: Yes, I plan to optimize `Status new_iterator(ColumnIterator** iterator)` in next PR, thanks!

File: be/src/storage/rowset/vectorized/segment_iterator.cpp [hunk 5]
```diff
@@ -1311,7 +1309,11 @@ Status SegmentIterator::_apply_bitmap_index() {
 
     for (size_t i = 0; i < bitmap_columns.size(); i++) {
         Roaring roaring;
-        BitmapIndexIterator* bitmap_iter = _bitmap_index_iterators[bitmap_columns[i]];
+        auto bitmap_iter_ptr = FindOrNull(_bitmap_index_iterators, i);
```
Comment (sduzh) [inline be/src/storage/rowset/vectorized/segment_iterator.cpp:1312]:
```suggestion
        auto bitmap_iter_ptr = FindOrNull(_bitmap_index_iterators, bitmap_columns[i]);
```

**Review by sduzh — APPROVED** [review]

**Comment by sduzh** [comment]
LGTM, PTAL @imay

**Pull request closed without merging**